source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1175,10 +1176,7 @@ bool ContextImpl::verifyCertChain(X509& leaf_cert, STACK_OF(X509) & intermediate
   int res = cert_validator_->doVerifyCertChain(ctx.get(), nullptr, leaf_cert, nullptr);
   // If |SSL_VERIFY_NONE|, the error is non-fatal, but we keep the error details.
   if (res <= 0 && SSL_CTX_get_verify_mode(ssl_ctx) != SSL_VERIFY_NONE) {
-    const int n = X509_STORE_CTX_get_error(ctx.get());
-    const int depth = X509_STORE_CTX_get_error_depth(ctx.get());
-    error_details = absl::StrCat("X509_verify_cert: certificate verification error at depth ",
-                                 depth, ": ", X509_verify_cert_error_string(n));
+    error_details = DefaultCertValidator::extractX509VerificationErrorInfo(ctx.get());
```

Comment:
Yes. That's where I originally moved this code from. However I think this `verifyCertChain` is called for quic stack, they are doing more message appending in that call site, https://github.com/envoyproxy/envoy/blob/110559f09c62175aa3cefb742ae2c66e953b6597/source/common/quic/envoy_quic_proof_verifier.cc#L53 

Do you mean pass in a string pointer to the `doVerifyCertChain` function, and have all validator implementation to fill in that? I tend to agree, but that's a larger refactoring, was thinking to get green light before prorceed.

Reply:
> However I think this verifyCertChain is called for quic stack,

oh right! sorry about that.

So I suggest we put `extractX509VerificationErrorInfo` in Tls::Utility namespace (https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/utility.h) rather than defining it as a static method of DefaultCertValidator.

Reply:
Done.